python/ray/tests/test_basic.py

```diff
@@ -3028,11 +3028,47 @@ def test_shutdown_disconnect_global_state():
     assert str(e.value).endswith("ray.init has been called.")
 
 
+@pytest.mark.parametrize(
+    "ray_start_object_store_memory", [150 * 1024 * 1024], indirect=True)
+def test_put_pins_object(ray_start_object_store_memory):
+    x_id = ray.put("HI")
+    x_copy = ray.ObjectID(x_id.binary())
+    assert ray.get(x_copy) == "HI"
+
+    # x cannot be evicted since x_id pins it
+    for _ in range(10):
+        ray.put(np.zeros(10 * 1024 * 1024))
+    assert ray.get(x_id) == "HI"
+    assert ray.get(x_copy) == "HI"
+
+    # now it can be evicted since x_id pins it but x_copy does not
+    del x_id
+    for _ in range(10):
+        ray.put(np.zeros(10 * 1024 * 1024))
+    with pytest.raises(ray.exceptions.UnreconstructableError):
+        ray.get(x_copy)
+
+    # weakref put
+    y_id = ray.put("HI", weakref=True)
```

Comment:
FYI @devin-petersohn after this PR if you use ray.put you might want to change modin to use weakref puts internally. Otherwise the puts will never be evicted while you have a ref. Not sure if this is an issue for you.

Reply:
That might actually be the right behavior for modin.

Reply:
Hm I imagine you probably don't want puts getting evicted within a single query. But then you want old pandas dataframes to get evicted. So maybe some weak/soft reference is warranted but at a different level.

Reply:
Thanks I will give it a try.

99% of the time we are not using puts, data is ingested from a task.